### PR TITLE
Export with options

### PR DESCRIPTION
**Pull request by fxha**
| Q                | A
| ---------------- | ---
| Bug fix?         | yes
| New feature?     | yes
| Fixed tickets    | #349, #354, #1479
| License          | MIT

### Description

- Added export/print settings
- Added custom and user-defined themes for exporting/printing
- Fixed code block overflow in PDF
- Known issues:
  - The blockquote border is rendered over header/footer when it's over multiple pages.
  - Emojis may be black when exporting on Linux if `sans-serif` or `serif` is defined because the system fallback font is used instead colored emojis. `Noto Color Emoji` cannot overwrite this.
  - Adding Noto Color Emoji to font fallback list result in files with multiple MB (because the font is included in PDF) and the emojis look weird due to the point above.
- Improvements: Print in another window for better performance - you could export/print and write at the same time.
- To-do in another PR: Documentation

### CODE REVIEW

**Comment by Jocs** [comment]
> Print in another window for better performance - you could export/print and write at the same time.

Good idea, maybe we can use a headless Electron window for export/print.

**Comment by Jocs** [comment]
Nice features and some UI suggestion:

1. Put `Info` tab content bellow `Export Options`, and no need to give it an individual tab. Because it has no settings of export PDF/HTML.
屏幕快照 2019-10-23 上午11 28 30

2. Narrow the vertical scroll bar, maybe 5px is ok? or just hide it?

屏幕快照 2019-10-23 上午11 29 44

File: src/muya/lib/utils/exportHtml.js [hunk 1]
```diff
@@ -222,12 +249,103 @@ class ExportHtml {
...
+  return `<div class="page-footer ${footerClass}">
+  <div class="hf-container">
+    <div class="footer-content-left">${left}</div>
+    <div class="footer-content">${center}</div>
+    <div class="footer-content-right">${right}</div>
+  </div>
+</div>`
+}
+
+const MARKDOWN_ARTICLE = html => {
```
Comment (Jocs) [inline src/muya/lib/utils/exportHtml.js]: Why do you use Uppercase name of this function?

Comment (fxha) [reply src/muya/lib/utils/exportHtml.js]: It was const before :smile:

**Comment by fxha** [comment]
>  1. Put `Info` tab content bellow `Export Options`, and no need to give it an individual tab. Because it has no settings of export PDF/HTML.
>    屏幕快照 2019-10-23 上午11 28 30

I prefer to have a separate start page and not the options on the first page. @Jocs  What do you think about that?

> Narrow the vertical scroll bar, maybe 5px is ok? or just hide it?

I'll narrow the bar a bit.

File: src/muya/lib/utils/exportHtml.js (hunk 1 above)
Comment (Jocs) [reply src/muya/lib/utils/exportHtml.js]: It seems strange, we only use UPPERCASE name in vuex action now.

**Comment by Jocs** [comment]
>I prefer to have a separate start page and not the options on the first page.

What do you mean `a separate start page`?

**Comment by fxha** [comment]
> > I prefer to have a separate start page and not the options on the first page.
> 
> What do you mean `a separate start page`?

A separate tab like the `Info` tab in the image above - keep it as it is.

**Comment by Jocs** [comment]
截屏2019-10-24下午11 05 17

We can borrow the UI from Chrome print setting modal in the future, and show a preview page on the left.

**Comment by fxha** [comment]
That's a good idea and also brings better export/print performance because it's doesn't block the editor. BTW page number are currently not supported because it's not possible with pure CSS or current Electron.

**Comment by Jocs** [comment]
>That's a good idea and also brings better export/print performance because it's doesn't block the editor.

Yes, you mean a separate window, it will bring better performance.

**Comment by Jocs** [comment]
@fxha I found that I can not set the `margin/padding` in the export setting now, and `header/footer` also can not be setted?

**Comment by fxha** [comment]
>  I found that I can not set the `margin/padding` in the export setting now, and `header/footer` also can not be setted?

Could you please explain the issue or provide a gif? Any exceptions or so?

**Comment by Jocs** [comment]
>Could you please explain the issue or provide a gif? Any exceptions or so?

It works well now, I export it in the morning, the `Page` and `Header & Footer` seems disappeared, but I can not reproduce it again.

**Pull request merged**